client/src/js/services/SearchFilterFormatService.js

```diff
@@ -0,0 +1,141 @@
+angular.module('bhima.services')
+.service('SearchFilterFormatService', SearchFilterFormatService);
+
+// dependencies injection 
+SearchFilterFormatService.$inject = [];
+
+/**
+ * SearchFilterFormatService
+ * @description This service help to format filters which are displayed for a search 
+ */
+function SearchFilterFormatService() {
+
+    var service = this;
+
+    // expose the service 
+    service.formatFilter = formatFilter;
+    service.formatDisplayNames = formatDisplayNames;
+    service.onRemoveFilter = onRemoveFilter;
+    service.clearFilters = clearFilters;
+
+    /**
+     * @function formatFilter
+     * @description filter and format properties of an object for being processed by formatFilterValues
+     * @param {object} element the object which contains all filters and their values. Ex. { reference: 'TPA1', is_cancelled: 1 }
+     * @param {boolean} WITH_NULL_VALUES a boolean flag for returning also properties which have null values 
+     * @return {object} {identifiers: ..., display: ...} returns identifiers for search query and display: values to display in filter 
+     */
+    function formatFilter(element, WITH_NULL_VALUES) {
+      var queryParam = formatFilterParameters(element, true);
+      var params = formatFilterValues(queryParam);
+      return params;
+    }
+
+    /**
+     * @function formatFilterParameters
+     * @description filter and format properties of an object for being processed by formatFilterValues
+     * @param {object} element the object which contains all filters and their values. Ex. { reference: 'TPA1', is_cancelled: 1 }
+     * @param {boolean} WITH_NULL_VALUES a boolean flag for returning also properties which have null values 
+     */
+    function formatFilterParameters(element, WITH_NULL_VALUES) {
+        var out = {};
+        for (var i in element) {
+            if (WITH_NULL_VALUES) {
+            out[i] = element[i];
+            } else if (element[i]) {
+            out[i] = element[i];
+            }
+        }
+        return out;
+    }
+
+  /**
+   * @function formatFilterValues
+   * @description identifier and display value
+   * @param {object} formatedFilters a returned value of formatFilterParameters
+   * @return {object} fomatedValues { identifiers: {}, display: {} }
+   */
+  function formatFilterValues(formatedFilters) {
+    var out = { identifiers: {}, display: {} };
+    for (var key in formatedFilters) {
+
+      if (!formatedFilters.hasOwnProperty(key)) { continue; }
+
+      // get identifiers
+      out.identifiers[key] = typeof(formatedFilters[key]) === 'object' ?
```

Comment:
I did try this suggestion in a `angular.forEach` loop but it doesn't work, so i just create `getFormattedFilterIdentifier` function and i use it in a normal `for(...)` loop.